test_suite/benchmark_test.py

```diff
@@ -0,0 +1,91 @@
+import sys
+import unittest
+import subprocess
+import os
+import glob
+
+import testenv
+
+class BenchmarkTest(unittest.TestCase):
+
+  class BenchmarkResults:
+    def __init__(self, title, resource_usage):
+      self.ru = resource_usage
+      self._title = title
+
+    @property
+    def title(self):
+      return self._title
+
+    @title.setter
+    def title(self, title):
+        self.title_ = title
+
+    @property
+    def time(self):
+      return "{0:.2f}".format((self.ru.ru_utime + self.ru.ru_stime) * 1000.0)
+
+    @property
+    def memory(self):
+      if sys.platform == 'darwin':
+        return "{0:.2f}".format(self.ru.ru_maxrss / 1024)
+      return "{0:.2f}".format(self.ru.ru_maxrss)
+
+
+  def benchmark(self, f, binary):
+    """Execute Starlark file, return stderr."""
+    proc = subprocess.Popen(
+        [glob.glob(testenv.STARLARK_BINARY_PATH[binary])[0], f],
```

Comment:
What type of benchmarks do you have in mind?
If these are micro-benchmarks, then the JVM overhead will dominate the Starlark runtime performance.

Reply:
Not micro-benchmarks, but operations with a large number of iterations (10^8 for example) known complexity (sorting, hashing, string algorithms, ...). I know that JVM has a large overhead, but with such benchmarks, we can still get useful info. For example, having benchmarks for operations than should run in O(n) with 10^8 iterations, or O(n^2) with 10^4 iterations and so on, and comparing benchmarks with the same complexity with each other can help identify if some slower operation has an area for improvement. The same goes for comparing memory consumption between implementations. Values of time and memory may not have a meaning by themselves but comparing those values with other benchmarks may do...

Reply:
I discussed with Laurent offline today. If in the benchmarks, you make sure that you run enough iterations so that the JVM startup doesn't dominate the overall runtime of the test (say a couple of seconds), then it's fine.

Let me know if you already have a list of benchmarks in mind.

Reply:
I added some benchmarks for the basic functions in Starlark. Please let me know what you think.
Other benchmarks I have in mind are, as I previously mentioned, some general algorithms and operations on data structures. Examples are benchmarks for graph algorithms, binary search, heaps, trees, tries, etc. I think they can be useful where if there are some performance issues in Starlark then these benchmarks would help detect them by producing results contradicting the expected complexities of these benchmarks.

Reply:
I wonder whether those general algorithms are a good indicator of starlark performance when it comes to its use in bazel. But it's a starting point, so feel free to go ahead.